internal/runtime/catalog/catalog.go

```diff
@@ -0,0 +1,333 @@
+		catalogName: naming.GetNameFromCallsite("sigs.k8s.io/cluster-api/internal/runtime/catalog/catalog.go"),
+	}
+}
+
+// AddHook adds a HookFunc and its request and response types with the gv GroupVersion.
+// The passed in hookFunc must have the following type: func(*RequestType,*ResponseType)
+// The name of the func becomes the "hook" in GroupVersionHook.
+// GroupVersion must not have empty fields.
+func (c *Catalog) AddHook(gv schema.GroupVersion, hookFunc Hook, hookMeta *HookMeta) {
+	// Validate gv.Group and gv.Version are not empty.
+	if gv.Group == "" {
+		panic("group must not be empty")
+	}
+	if gv.Version == "" {
+		panic("version must not be empty")
+	}
+
+	// Validate that hookFunc is a func.
+	t := reflect.TypeOf(hookFunc)
+	if t.Kind() != reflect.Func {
+		panic("hook must be a func")
+	}
+	if t.NumIn() != 2 {
+		panic("hook must have two input parameter: *RequestType, *ResponseType")
+	}
+	if t.NumOut() != 0 {
+		panic("hook must have no output parameter")
+	}
+
+	// Create request and response objects based on the input types.
+	request, ok := reflect.New(t.In(0).Elem()).Interface().(runtime.Object)
+	if !ok {
+		panic("hook request (first parameter) must be a runtime.Object")
+	}
+	response, ok := reflect.New(t.In(1).Elem()).Interface().(runtime.Object)
+	if !ok {
+		panic("hook response (second parameter) must be a runtime.Object")
+	}
+
+	// Calculate the hook name based on the func name.
+	hookFuncName := goruntime.FuncForPC(reflect.ValueOf(hookFunc).Pointer()).Name()
+	hookName := hookFuncName[strings.LastIndex(hookFuncName, ".")+1:]
+
+	gvh := GroupVersionHook{
+		Group:   gv.Group,
+		Version: gv.Version,
+		Hook:    hookName,
+	}
+
+	// Validate that the GVH is not already registered with another type.
+	if oldT, found := c.gvhToType[gvh]; found && oldT != t {
+		panic(fmt.Sprintf("double registration of different type for %v: old=%v.%v, new=%v.%v in catalog %q", gvh, oldT.PkgPath(), oldT.Name(), t.PkgPath(), t.Name(), c.catalogName))
+	}
+
+	// Add GVH <=> RuntimeHook mappings.
+	c.gvhToType[gvh] = t
+	c.typeToGVH[t] = gvh
+
+	// Add Request and Response types to scheme.
+	c.scheme.AddKnownTypes(gv, request)
+	c.scheme.AddKnownTypes(gv, response)
+
+	// Create a hook descriptor and store it in the GVH => Descriptor map.
+	requestGVK, err := c.GroupVersionKind(request)
+	if err != nil {
+		panic(fmt.Sprintf("failed to get GVK for request %T: %v", request, err))
+	}
+	responseGVK, err := c.GroupVersionKind(response)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to get GVK for response %T: %v", request, err))
+	}
+	c.gvhToHookDescriptor[gvh] = hookDescriptor{
+		metadata: hookMeta,
+		request:  requestGVK,
+		response: responseGVK,
+	}
+}
+
+// AddOpenAPIDefinitions adds an OpenAPIDefinitionsGetter with the gv GroupVersion.
+func (c *Catalog) AddOpenAPIDefinitions(gv schema.GroupVersion, getter OpenAPIDefinitionsGetter) {
```

Comment:
What about adding some validation on the input parameters?

Reply:
I am note sure what other kinds of validations can we add here. 

Note: empty `gv` is considered valid here. We use that to capture the OpenAPI definitions of all the vendored types.

Reply:
We can also defer this to the openapi gen PR, it's probably easier to discuss there